### PR TITLE
[TTS] [黑客松]Add JETS

### LINKED ISSUE #2773: [TTS] JETS -> E2E FastSpeech2 + HiFiGAN
E2E FastSpeech2 + HiFiGAN

Paper: [JETS: Jointly Training FastSpeech2 and HiFi-GAN for End to End Text to
Speech](https://arxiv.org/pdf/2203.16852.pdf)

Also check for more information about E2E-training TTS:
- https://github.com/PaddlePaddle/PaddleSpeech/issues/1699

### PR DESCRIPTION
**Pull request by ljhzxc**
### PR types
New features

### PR changes
新增了 jets 模型

### Describe
新增 jets 模型，以及 example 的 csmcs ，包括 文档，动态图训练推理，动态图转静态图，推理。
复现过程中存在问题：
Paddle 的 ctcloss 和 torch 的 ctcloss 的接口是不一样的。
torch 的 [ctcloss](https://pytorch.org/docs/1.10/generated/torch.nn.functional.ctc_loss.html?highlight=ctc#torch.nn.functional.ctc_loss)
paddle 的 [ctcloss](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/CTCLoss_cn.html):
虽然 paddle 的 [nn.nn.functional.ctc_loss](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/functional/ctc_loss_cn.html) 文档中说明输入是 log_probs, 但似乎不是这样。源码中也是调用 warpctc，而 warpctc 输入是 logits。 

综上所述：torch 输入是 log_probs，但是 paddle 输入是 logits。
image
模型中需要先对 log_probs 做 β伯努利，但是 paddle 的 ctc 由于输入是 logits，不能实现这个步骤，所以不能和原版模型对齐。这导致了自动对齐模块不能实现，因此最终模型采用了 mfa 的对齐结果进行训练。

fix https://github.com/PaddlePaddle/PaddleSpeech/issues/2773

### CODE REVIEW

**Comment by lizezheng** [comment]
请问有 csmsc 训练完成的 jets 模型，合成的语音样例么？

File: examples/csmsc/jets/local/inference.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+stop_stage=0
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    python3 ${BIN_DIR}/inference.py \
+        --inference_dir=${train_output_path}/inference \
+        --am=jets_csmsc \
+        --text=${BIN_DIR}/../sentences.txt \
+        --output_dir=${train_output_path}/pd_infer_out \
+        --phones_dict=dump/phone_id_map.txt
+fi
```
Comment (lym0302) [inline examples/csmsc/jets/local/inference.sh:15]: local 下的所有文件的权限用chmod 加一下，否则 bash 用不了

File: paddlespeech/t2s/exps/jets/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    parser.add_argument(
+        "--write_metadata_method",
+        default="w",
+        type=str,
+        choices=["w", "a"],
+        help="How the metadata.jsonl file is written.")
+
+    parser.add_argument(
+        "--token_average",
+        action="store_true",
```
Comment (lym0302) [inline paddlespeech/t2s/exps/jets/preprocess.py]: 建议使用str2bool 更直观

File: examples/csmsc/jets/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
+# JETS with CSMSC
+This example contains code used to train a [JETS](https://arxiv.org/abs/2203.16852v1) model with [Chinese Standard Mandarin Speech Copus](https://www.data-baker.com/open_source.html).
+
+## Dataset
+### Download and Extract
+Download CSMSC from it's [Official Website](https://test.data-baker.com/data/index/source).
+
+### Get MFA Result and Extract
+We use [MFA](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner) to get phonemes adn durations for JETS.
```
Comment (lym0302) [inline examples/csmsc/jets/README.md]: adn 拼写错误？

File: paddlespeech/t2s/exps/jets/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+            token_average=args.token_average)
+    if dev_wav_files:
+        process_sentences(
+            config=config,
+            fps=dev_wav_files,
+            sentences=sentences,
+            output_dir=dev_dump_dir,
+            mel_extractor=mel_extractor,
+            pitch_extractor=pitch_extractor,
+            energy_extractor=energy_extractor,
```
Comment (lym0302) [inline paddlespeech/t2s/exps/jets/preprocess.py:428]: 下面加上这个：nprocs=args.num_cpu,

File: paddlespeech/t2s/exps/syn_utils.py [hunk 5]
```diff
@@ -528,6 +528,18 @@ def am_to_static(am_inference,
                 InputSpec([-1], dtype=paddle.int64),  # is_slur
             ])
 
+    elif am_name == 'jets':
```
Comment (lym0302) [inline paddlespeech/t2s/exps/syn_utils.py]: 这里是不是可以和 vits 放一起？

File: paddlespeech/t2s/exps/jets/preprocess.py (hunk 2 above)
Comment (ljhzxc) [reply paddlespeech/t2s/exps/jets/preprocess.py]: done

Comment (ljhzxc) [reply paddlespeech/t2s/exps/jets/preprocess.py]: ok

File: examples/csmsc/jets/README.md (hunk 3 above)
Comment (ljhzxc) [reply examples/csmsc/jets/README.md]: 修复了

File: paddlespeech/t2s/exps/jets/preprocess.py (hunk 4 above)
Comment (ljhzxc) [reply paddlespeech/t2s/exps/jets/preprocess.py:428]: ok

File: paddlespeech/t2s/exps/syn_utils.py (hunk 5 above)
Comment (ljhzxc) [reply paddlespeech/t2s/exps/syn_utils.py]: 可以

**Review by yt605155624 — DISMISSED** [review]
LGTM

**Review by yt605155624 — APPROVED** [review]

**Comment by zh794390558** [comment]
ctc_loss的输入可以是log_softmax的，可以验证下。

**Comment by ljhzxc** [comment]
image
因为底层的 warpctc 当中存在了 softmax，所以只能用 logits 作为输入，这里文档里的输入应该写错了
